### PR TITLE
Use CachedFactoryContext

### LINKED ISSUE #45840: `RequestDelegateMetadataResult.CachedFactoryContext` is unused, resulting in unnecessary work
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I was looking through some of the code related to `RequestDelegateFactory` and I noticed that `RequestDelegateMetadataResult.CachedFactoryContext` doesn't appear to be used. I'm not sure whether this is a bug or by design, but it seems like not using it results in a lot of duplicated work building `Expression`s? Especially as the documentation for `CachedFactoryContext` says:

```csharp
// This internal cached context avoids redoing unnecessary reflection in Create that was already done in InferMetadata.
// InferMetadata currently does more work than it needs to building up expression trees, but the expectation is that InferMetadata will usually be followed by Create.
internal RequestDelegateFactoryContext? CachedFactoryContext { get; set; }
```

### Expected Behavior

I would expect the `CachedFactoryContext` [to be set in `InferMetadata`](https://github.com/dotnet/aspnetcore/blob/45fcfb75dd7ca6c548200dde4f7374a9a451039b/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L135), e.g something like this:

```diff
public static RequestDelegateMetadataResult InferMetadata(MethodInfo methodInfo, RequestDelegateFactoryOptions? options = null)
{
    var factoryContext = CreateFactoryContext(options);
    factoryContext.ArgumentExpressions = CreateArgumentsAndInferMetadata(methodInfo, factoryContext);

    return new RequestDelegateMetadataResult
    {
        EndpointMetadata = AsReadOnlyList(factoryContext.EndpointBuilder.Metadata),
+       CachedFactoryContext = factoryContext,
    };
}
```

### Steps To Reproduce

I simply created a new minimal API app, debugged into `RequestDelegateFactory`, and placed a breakpoint in `CreateArguments()`. It's called both in `InferMetadata()` and `Create()`, which `CachedFactoryContext` is meant to avoid (if I understand correctly!)

### Exceptions (if any)

_No response_

### .NET Version

7.0.101

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by andrewlock**
# Fix `CachedFactoryContext` is unused

Stores the `RequestDelegateFactoryContext` created in `RequestDelegateFactory.InferMetadata()` on `RequestDelegateMetadataResult.CachedFactoryContext` so it can be reused in `RequestDelegateFactory.Create()`.

## Description

This was always the _expected_ behaviour, as caching the factory context prevents a lot of re-work building `Expression` trees for the handlers parameters.

Without the caching, the behaviour was still correct (as evidenced by the unit tests that assume the caching was taking place), it would just have an impact on startup time.

Fixes #45840

### CODE REVIEW

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 1]
```diff
@@ -6935,6 +6935,23 @@ public void InferMetadata_PopulatesAcceptsMetadata_WhenReadFromForm()
...
+        var options = new RequestDelegateFactoryOptions
+        {
+            EndpointBuilder = CreateEndpointBuilder(),
+        };
+
+        // Act
+        var metadataResult = RequestDelegateFactory.InferMetadata(@delegate.Method, options);
+
+        // Assert
+        Assert.NotNull(metadataResult.CachedContext);
```
Comment (halter73) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:6952]:
```suggestion
        Assert.NotNull(metadataResult.CachedFactoryContext);
```

**Review by halter73 — APPROVED** [review]

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs (hunk 1 above)
Comment (andrewlock) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs]: Well... that's awkward 🤣

**Comment by halter73** [comment]
Thanks @andrewlock!

**Pull request merged**